File: tests/loop/shift/loop_shift_basic.ml
open Optitrust
open Prelude

let _ = Flags.check_validity := true
let _ = Flags.recompute_resources_between_steps := true

let _ = Run.script_cpp(fun _ ->
  !! Loop_basic.shift_range "i2" (ShiftBy (trm_int 2)) [cFunBody "main"; cFor "i"];
  !! Loop_basic.shift_range "j2" StartAtZero [cFunBody "main"; cFor "j"];
  (* FIXME:
  !! Loop_basic.shift_range "k2" (ShiftBy (expr "shift")) [cFunBody "main"; cFor "k"];
  !! Loop_basic.shift_range "l2" (ShiftBy (expr "shift")) [cFunBody "main"; cFor "l"];
  !! Loop_basic.shift_range "m2" (StopAt (expr "N")) [cFunBody "main"; cFor "m"]; *)
  !! Loop_basic.shift_range "m3" (StartAt (trm_int 4)) [cFunBody "main"; cFor "m"]; (* "m2" *)

  !! Loop_basic.shift_range "j" (ShiftBy (trm_int 2)) [cFunBody "excl_array"; cFor "i"];
)

File: tests/loop/shift/loop_shift_basic.cpp
#include <optitrust.h>

int main(){
  __pure();

  int x = 0;
  const int st = 0;
  const int N = 10;

  for (int i = 0; i < 10; i++){
    __sequentially_modifies("&x ~> Cell");
    x += i;
  }

  for (int j = st; j < st+N; j++){
    __sequentially_modifies("&x ~> Cell");
    x += j;
  }

  int shift = 5;
  for (int k = 0; k < N; k++){
    __sequentially_modifies("&x ~> Cell");
    x += k;
  }
/* FIXME:
  for (int l = N; l > 0; l--){
    __sequentially_modifies("&x ~> Cell");
    x += l;
  }
*/
  for (int m = 2; m < N-2; m++) {
    __sequentially_modifies("&x ~> Cell");
    x += m;
  }
}

void excl_array(int* t, int n) {
  __modifies("t ~> Matrix1(n)");

  for (int i = 0; i < n; i++){
    __modifies("&t[MINDEX1(n, i)] ~> Cell");
    t[MINDEX1(n, i)] += i;
  }
}
